File: cpp/data_structures/dsu.cpp
/**
 * Author: concl
 * Status: Somewhat tested
 */

#include <bits/stdc++.h>

using namespace std;

class DSU {
public:
    vector<int> dsu;
    vector<int> sizes;

    DSU(int n) {
        dsu.resize(n);
        sizes.resize(n, 1);
        for (int i = 0; i < n; i++) {
            dsu[i] = i;
        }
    }

    int find(int node) {
        if (dsu[node] == node) {
            return node;
        } else {
            return dsu[node] = find(dsu[node]);
        }
    }

    void union_sets(int a, int b) {
        int x = find(a);
        int y = find(b);

        if (x == y) {
            return;
        } else if (sizes[x] < sizes[y]) {
            swap(x, y);
        }

        dsu[y] = x;
        sizes[x] += sizes[y];
    }
};

// Functional implementation of disjoint set union (union find) data structure
int find(vector<int> &dsu, int node) {

    if (dsu[node] == node) {
        return node;
    } else {
        return dsu[node] = find(dsu, dsu[node]);
    }
}

void union_sets(vector<int> &dsu, vector<int> &sizes, int a, int b) {

    int x = find(dsu, a);
    int y = find(dsu, b);

    if (x == y) {
        return;
    } else if (sizes[x] < sizes[y]) {
        swap(x, y);
    }

    dsu[y] = x;
    sizes[x] += sizes[y];
}

File: python/data_structures/dsu.py

class DSU:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.size = [1] * n
        
    
    def find(self, a: int) -> int:
        if self.parent[a] != a:
            self.parent[a] = self.find(self.parent[a])
        return self.parent[a]

    def union_sets(self, a: int, b: int):
        a = self.find(a)
        b = self.find(b)
        if a == b:
            return
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.parent[b] = a
        self.size[a] += self.size[b]

# Procedural implementation
def find(parent, a):
    if parent[a] != a:
        parent[a] = find(parent, parent[a])
    return parent[a]

def union_sets(parent, sizes, a, b):

    a = find(parent, a)
    b = find(parent, b)
    if a == b:
        return
    if sizes[a] < sizes[b]:
        a, b = b, a
    parent[b] = a
    sizes[a] += sizes[b]


